### PR TITLE
JavaScript Suite SDL Manager

### PR DESCRIPTION
**Pull request by crokita**
In a similar vein to [this proposal](https://github.com/smartdevicelink/sdl_evolution/blob/master/proposals/0171-android-manager-apis.md), the goal is to introduce a manager layer to the JavaScript Suite, closely matching the Java Suite and iOS manager APIs. These manager classes are designed to abstract out complex flows and states for common tasks, such as file handling and displaying information to the screen.

### CODE REVIEW

**Comment by crokita** [comment]
@theresalech ready for review

File: proposals/nnnn-sdl-javascript-manager-layer.md [hunk 1]
```diff
@@ -0,0 +1,96 @@
+# JavaScript Suite SDL Manager
+
+* Proposal: [SDL-NNNN](NNNN-js-suite-managers.md)
```
Comment (theresalech) [inline proposals/nnnn-sdl-javascript-manager-layer.md]:
```suggestion
* Proposal: [SDL-NNNN](NNNN-sdl-javascript-manager-layer.md)
```

File: proposals/nnnn-sdl-javascript-manager-layer.md [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+* Author: [Chris Rokita](https://github.com/crokita)
+* Status: **Awaiting review**
+* Impacted Platforms: JavaScript Suite
+
+## Introduction
+
+In a similar vein to [this proposal](https://github.com/smartdevicelink/sdl_evolution/blob/master/proposals/0171-android-manager-apis.md), the goal is to introduce a manager layer to the JavaScript suite, closely matching the Android and iOS manager APIs. These manager classes are designed to abstract out complex flows and state for common tasks, such as file handling and displaying information to the screen.
```
Comment (theresalech) [inline proposals/nnnn-sdl-javascript-manager-layer.md]:
```suggestion
In a similar vein to [this proposal](https://github.com/smartdevicelink/sdl_evolution/blob/master/proposals/0171-android-manager-apis.md), the goal is to introduce a manager layer to the JavaScript Suite, closely matching the Java Suite and iOS manager APIs. These manager classes are designed to abstract out complex flows and states for common tasks, such as file handling and displaying information to the screen. 
```
Please make sure you update the PR description with these changes as well, thanks!

File: proposals/nnnn-sdl-javascript-manager-layer.md [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+
+## Motivation
+
+Boilerplate code that developers will have to write in order to effectively use the library's features is time that doesn't need to be spent. In addition, having developers write their own abstractions to common tasks introduces the possibility of bugs that will slow down the process of using the library. Having a set of built in classes that are designed to remove setup work means more time saved for every developer that uses the project and less error-prone implementations for future apps.
+
+## Proposed solution
+
+A set of submanager classes will be created, with each one designated to a particular responsibility to simplify using the library. These submanagers will be controlled by an encompassing class named `SdlManager`, which is for the developer's convenience. These classes will match closely to existing manager layers to allow for easy adoption for those already familiar with the Android or iOS libraries. In addition, unique features of JavaScript can be utilized to further ease the pain of dealing with asynchronous logic, such as using Promises and the yielding async/await syntax.
```
Comment (theresalech) [inline proposals/nnnn-sdl-javascript-manager-layer.md]:
```suggestion
A set of submanager classes will be created, with each one designated to a particular responsibility to simplify using the library. These submanagers will be controlled by an encompassing class named `SdlManager`, which is for the developer's convenience. These classes will match closely to existing manager layers to allow for easy adoption for those already familiar with the Java Suite or iOS libraries. In addition, unique features of JavaScript can be utilized to further ease the pain of dealing with asynchronous logic, such as using Promises and the yielding async/await syntax. 
```

File: proposals/nnnn-sdl-javascript-manager-layer.md [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+
+## Proposed solution
+
+A set of submanager classes will be created, with each one designated to a particular responsibility to simplify using the library. These submanagers will be controlled by an encompassing class named `SdlManager`, which is for the developer's convenience. These classes will match closely to existing manager layers to allow for easy adoption for those already familiar with the Android or iOS libraries. In addition, unique features of JavaScript can be utilized to further ease the pain of dealing with asynchronous logic, such as using Promises and the yielding async/await syntax. 
+
+### Managers
+
+Specifically, these are the following manager classes that existing libraries already use, and thus will need to be implemented in the JavaScript suite:
+* BaseSubManager (The class that submanagers will extend)
+* SdlManager (Publically-facing manager that allows access to the submanagers below)
```
Comment (theresalech) [inline proposals/nnnn-sdl-javascript-manager-layer.md]:
```suggestion
* SdlManager (Publicly-facing manager that allows access to the submanagers below)
```

File: proposals/nnnn-sdl-javascript-manager-layer.md [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+
+A lot of code will need to be written and tested, placing the burden of boilerplate on the project maintainers instead of the developers. But, the time saved in the long run for future developers will be much larger than the time taken to write the managers out now. 
+
+## Impact on existing code
+
+The code changes required are additive and will not impact existing code.
+
+## Alternatives considered
+
+Deviating from the existing manager layers' APIs is an option, but it comes at the cost of being a different set of APIs, and so a developer migrating from the Android or iOS libraries would have to learn and understand the manager layer all over again.
```
Comment (theresalech) [inline proposals/nnnn-sdl-javascript-manager-layer.md]:
```suggestion
Deviating from the existing manager layers' APIs is an option, but it comes at the cost of being a different set of APIs, and so a developer migrating from the Java Suite or iOS libraries would have to learn and understand the manager layer all over again.
```

**Comment by crokita** [comment]
@theresalech I have made the requested changes

**Pull request merged**